test/cpp/util/channelz_sampler.cc

```diff
@@ -0,0 +1,512 @@
+        std::cout << so_id << " ";
+        grpc::channelz::v1::Socket so = GetSocketRPC(so_id);
+        if (CheckID(so_id)) {
+          all_sockets.push_back(so);
+          StoreSocketInJson(so);
+        }
+      }
+    }
+    std::cout << std::endl;
+  }
+
+  // Set up the channelz sampler client
+  // Initialize json as an array
+  void Setup(std::string custom_credentials_type, std::string server_address) {
+    json = grpc_core::Json::Array();
+    grpc::ChannelArguments channel_args;
+    std::shared_ptr<grpc::ChannelCredentials> channel_creds =
+        grpc::testing::GetCredentialsProvider()->GetChannelCredentials(
+            custom_credentials_type, &channel_args);
+    std::shared_ptr<grpc::Channel> channel =
+        CreateChannel(server_address, channel_creds);
+    channelz_stub = grpc::channelz::v1::Channelz::NewStub(channel);
+  }
+
+  // Get all servers, keep querying until getting all
+  // Store servers for dumping data
+  // Need to check id repeating for servers
+  bool GetServersRPC() {
+    int64_t server_start_id = 0;
+    while (true) {
+      GetServersRequest get_server_request;
+      GetServersResponse get_server_response;
+      ClientContext get_server_context;
+      get_server_request.set_start_server_id(server_start_id);
+      Status status = channelz_stub->GetServers(
+          &get_server_context, get_server_request, &get_server_response);
+      if (!status.ok()) {
+        gpr_log(GPR_ERROR, "%s",
+                get_server_context.debug_error_string().c_str());
+        return false;
+      }
+      for (const auto& _server : get_server_response.server()) {
+        all_servers.push_back(_server);
+        StoreServerInJson(_server);
+      }
+      if (!get_server_response.end()) {
+        server_start_id = GetServerID(all_servers.back()) + 1;
+      } else {
+        break;
+      }
+    }
+    std::cout << "Number of servers = " << all_servers.size() << std::endl;
+    return true;
+  }
+
+  // Get sockets that belongs to servers
+  // Store sockets for dumping data
+  void GetSocketsOfServers() {
+    for (const auto& _server : all_servers) {
+      std::cout << "Server " << GetServerID(_server) << " listen_socket: ";
+      for (const auto& _socket : _server.listen_socket()) {
+        int64_t so_id = _socket.socket_id();
+        std::cout << so_id << " ";
+        if (CheckID(so_id)) {
+          grpc::channelz::v1::Socket so = GetSocketRPC(so_id);
+          all_sockets.push_back(so);
+          StoreSocketInJson(so);
+        }
+      }
+      std::cout << std::endl;
+    }
+  }
+
+  // Get all top channels, keep querying until getting all
+  // Store channels for dumping data
+  // No need to check id repeating for top channels
+  bool GetTopChannelsRPC() {
+    int64_t channel_start_id = 0;
+    while (true) {
+      GetTopChannelsRequest get_top_channels_request;
+      GetTopChannelsResponse get_top_channels_response;
+      ClientContext get_top_channels_context;
```

Comment:
nit: lets put (generous) deadlines on all of these RPCs, just so that the tool doesn't get "stuck" somehow in one of these RPCs.

We can do something like, `get_top_channels_context.set_deadline(grpc_timeout_seconds_to_deadline(20)`

Reply:
done